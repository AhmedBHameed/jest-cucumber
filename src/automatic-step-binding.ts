import { ParsedFeature } from './models';
import { matchSteps } from './validation/step-definition-validation';
import { StepsDefinitionCallbackFunction, defineFeature } from './feature-definition-creation';

const globalSteps: Array<{ stepMatcher: string | RegExp, stepFunction: () => any }> = [];

const registerStep = (stepMatcher: string | RegExp, stepFunction: () => any) => {
    globalSteps.push({ stepMatcher, stepFunction });
};

export const autoBindSteps = (features: ParsedFeature[], stepDefinitions: StepsDefinitionCallbackFunction[]) => {
    stepDefinitions.forEach((stepDefinitionCallback) => {
        stepDefinitionCallback({
            defineStep: registerStep,
            given: registerStep,
            when: registerStep,
            then: registerStep,
            and: registerStep,
            but: registerStep,
            pending: () => {
                // Nothing to do
            },
        });
    });

    features.forEach((feature) => {
        defineFeature(feature, (test) => {
            feature.scenarios.forEach((scenario) => {
                test(scenario.title, (options) => {
                    scenario.steps.forEach((step) => {
                        const matches = globalSteps
                            .filter((globalStep) => matchSteps(step.stepText, globalStep.stepMatcher));

                        if (matches.length === 1) {
                            const match = matches[0];

                            options.defineStep(match.stepMatcher, match.stepFunction);
                        } else if (matches.length === 0) {
                            // TODO: Log error
                        } else {
                            // TODO: Log error
                        }
                    });
                });
            });
        });
    });
};
