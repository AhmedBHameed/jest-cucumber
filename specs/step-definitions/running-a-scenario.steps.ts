import { loadFeature, defineFeature } from '../../src';
import { MockTestRunner } from '../utils/mock-test-runner/mock-test-runner';
import { wireUpMockFeature } from '../utils/wire-up-mock-scenario';
import { runningAScenarioFeature, runningAScenarioSteps, runningAScenarioStepsWithMissingStep } from '../test-data/running-a-scenario';

const feature = loadFeature('./specs/features/running-a-scenario.feature');

defineFeature(feature, (test) => {
    let mockTestRunner: MockTestRunner;

    beforeEach(() => {
        mockTestRunner = new MockTestRunner();
    });

    test('Running a scenario', ({ given, when, then }) => {
        let testPromise: Promise<any>;

        given('I have a scenario', () => {
            wireUpMockFeature(mockTestRunner, runningAScenarioFeature, runningAScenarioSteps);
        });

        when('I run it', () => {
            testPromise = mockTestRunner.execute();
        });

        then('it should run successfully', async () => {
            await testPromise;
        });
    });

    test('Scenario missing a step', ({ given, when, then }) => {
        let errorMessage: string;

        given('I have a scenario that is missing a step', () => {
            // Nothing to do here
        });

        when('I run it', () => {
            try {
                wireUpMockFeature(mockTestRunner, runningAScenarioFeature, runningAScenarioStepsWithMissingStep);
            } catch (err) {
                errorMessage = err.message;
            }
        });

        then('it should fail with a validation error / missing step code', () => {
            expect(errorMessage).toBeTruthy();
            expect(errorMessage).toMatchSnapshot();
        });
    });
});
