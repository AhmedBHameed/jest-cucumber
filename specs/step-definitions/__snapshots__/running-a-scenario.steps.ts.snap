// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running a scenario Scenario missing a step 1`] = `
"Scenario \\"Doing some stuff\\" has 3 step(s) in the feature file, but 2 step definition(s) defined. Try adding the following code:

test('Doing some stuff', ({ given, when, then }) => {
	given('I did some stuff', () => {

	});

	when('I do some stuff', () => {

	});

	then('I should have done some stuff', () => {

	});
});

Expected step #3 in scenario \\"Doing some stuff\\" to match \\"I should have done some stuff\\". Try adding the following code:

then('I should have done some stuff', () => {

});"
`;
